programs/building/loft.sc

```diff
@@ -0,0 +1,446 @@
+);
+_draw_surf_tick() -> (
+    if(global_loft:'draw',
+        _draw_surf(5);
+        schedule(4, '_draw_surf_tick')
+    )
+);
+_set_block(pos)->( //function for setting block and storing their preavious value in global_affected_block
+    if(block(pos)!=global_loft:'material',
+        if((block(pos)==global_loft:'replace')||(global_loft:'replace'==null),
+            existing_block=block(pos);
+	        global_loft:'affected_blocks'+=[pos,existing_block];
+            set(pos,global_loft:'material')
+        )
+    )
+);
+_leng_approx(a,b,c)->( //function that approximate the length of a piece of quadratic spline given its 3 control points
+    L1=0.5*(a-b);
+    L2=0.5*(b-c);
+    L3=0.5*(a+b)-0.5*(b+c);
+    sqrt((L1:0)^2+(L1:1)^2+(L1:2)^2)+
+    sqrt((L2:0)^2+(L2:1)^2+(L2:2)^2)+
+    sqrt((L3:0)^2+(L3:1)^2+(L3:2)^2)
+);
+_paste_patch(p00,p01,p10,p11,a0,a1,b0,b1,c)->( //function used to paste a patch of the surface
+    v=0;
+    lvmax=max(
+        _leng_approx(p00,a0,p01),
+        _leng_approx(b0,c,b1),
+        _leng_approx(p10,a1,p11)
+    );
+    lumax=max(
+        _leng_approx(p00,b0,p10),
+        _leng_approx(a0,c,a1),
+        _leng_approx(p01,b1,p11)
+    );
+    if(lvmax>0, dv=0.7/lvmax, dv=1);
+    if(lumax>0, du=0.7/lumax, du=1);
+    while(v<=1,1000,
+        u=0;
+        while(u<=1,1000,
+            point=(p00*(1-v)^2+2*a0*v*(1-v)+p01*v^2)*(1-u)^2+     //point of the surface s
+                (2*b0*(1-v)^2+4*c*v*(1-v)+2*b1*v^2)*u*(1-u)+
+                (p10*(1-v)^2+2*a1*v*(1-v)+p11*v^2)*u^2;
+            p=[round(get(point,0)),round(get(point,1)),round(get(point,2))];
+            schedule(0,'_set_block',p);
+            if((u<1)&&(u+du>1),
+                u=1,
+                u=u+du
+            )
+        );
+        if((v<1)&&(v+dv>1),
+            v=1,
+            v=v+dv
+        )
+    )
+);
+paste(Block,replace)->(
+    global_loft:'material'=Block;
+    global_loft:'replace'=replace;
+    m=length(global_loft:'cset');
+    n=length(global_loft:'cset':0);
+    if((n>1)&&(m>1),
+        c_for(i=0,i<n-1,i=i+1,
+            c_for(j=0,j<m-1,j=j+1,
+                _paste_patch(
+                    global_loft:'cset':j:i,
+                    global_loft:'cset':j:(i+1),
+                    global_loft:'cset':(j+1):i,
+                    global_loft:'cset':(j+1):(i+1),
+                    global_loft:'a':j:i,
+                    global_loft:'a':(j+1):i,
+                    global_loft:'b':j:i,
+                    global_loft:'b':j:(i+1),
+                    global_loft:'c':j:i
+                )
+            )
+        );
+        schedule(1,'_add_to_history');
+    )
+);
+_add_to_history()->(
+	if(length(global_loft:'affected_blocks')==0,return);
+	operation={'affected_position'->global_loft:'affected_blocks'};
+	global_loft:'history'+=operation;
+	global_loft:'affected_blocks'=[];
+    print('done');
+);
+undo()->(
+	if(length(global_loft:'history')==0,print('no mooves to undo');return);

```

Comment:
`return` in scarpet is not a keyword, it's a function, so, you'd need to add brackets: `return()`

Reply:
Thanks for telling me. I was using `return` as I learned in C but it work differently in scarpet